draft-ietf-jsonpath-base.md

```diff
@@ -569,207 +583,257 @@ of node.
+Nothing is selected from a value which is not a object.
 
+Array indexing via `element-index` is a way of selecting a particular array element using a zero-based index.
+For example, selector `[0]` selects the first and selector `[4]` the fifth element of a sufficiently filled array.
 
+A negative `element-index` counts from the array end.
+For example, selector `[-1]` selects the last and selector `[-2]` selects the last but one element of an array with at least two elements.
 
-#### Array Selector
 
-##### Syntax
-{: numbered="false" toc="exclude"}
+### Index Wild Card Selector
 
-An array selector selects zero or more elements of an array node.
-An array selector takes the form of an index, which selects at most one element,
-or a slice, which selects zero or more elements.
+#### Syntax
+{: unnumbered}
+
+The index wild card selector has the form `[*]`.
 
 ~~~~ abnf
-union-element =/ array-index / array-slice
+index-wild-selector    = "[" "*" "]"  ;  asterisk enclosed by brackets
 ~~~~
 
-An array index is an integer (in base 10).
+#### Semantics
+{: unnumbered}
 
-~~~~ abnf
-array-index = integer
+An index selector combined with a single asterisk is a wild card. It
+selects all member values of an object as well as all elements of an
+array.
+Applying the `index-wild-selector` to a primitive JSON value (such as
+a number, string, or true/false/null) selects no node.
 
-integer = ["-"] ("0" / (DIGIT1 *DIGIT))
-                            ; optional - followed by 0 or
-                            ; sequence of digits with no leading zero
-DIGIT1 = %x31-39            ; non-zero digit
-~~~~
+The `index-wild-selector` behaves identically to the `dot-wild-selector`.
+
+### Array Slice Selector
 
-Note: the syntax does not allow integers with leading zeros such as `01` and `-01`.
+#### Syntax
+{: unnumbered}
 
-An array slice consists of three optional integers (in base 10) separated by colons.
+The array slice selector has the form `[<start>:<end>:<step>]`.
+It selects elements starting at index `<start>`, ending at — but
+not including — `<end>`, while incrementing by `step`.
 
 ~~~~ abnf
-array-slice = [ start ] ws ":" ws [ end ]
-                   [ ws ":" ws [ step ] ]
-start = integer
-end = integer
-step = integer
+slice-selector = "[" slice-index "]"
+
+slice-index    = ws [start] ws ":" ws [end] [ws ":" ws [step] ws]
+
+start          = int       ; included in selection
+end            = int       ; not included in selection
+step           = int       ; default: 1
+
+ws             = *( %x20 / ; Space
+                    %x09 / ; Horizontal tab
+                    %x0A / ; Line feed or New line
+                    %x0D ) ; Carriage return
 ~~~~
 
-Note: the array slices `:` and `::` are both syntactically valid, as are `:2:2`, `2::2`, and `2:4:`.
+The `slice-selector` consists of three optional decimal integers separated by colons.
```

Comment:
@gregsdennis pointed out before that omitting the three integers always leaves behind _two_ separating colons when only one colon is valid. The ABNF is correct, but the description here contradicts it.

Reply:
Wait, the step is optional in the syntax, too.

Reply:
The ABNF syntax is correct (last time I looked!). It's just the above sentence which is misleading.